lib/userMemoryStore.js

```diff
@@ -0,0 +1,143 @@
+const ATTEMPT_DELAY_MS = 500;
+
+function findById(id) {
+  return _.find(users, {id: id});
+}
+function findByUsername(username) {
+  return _.find(users, {username: username});
+}
+
+function cloneAndCleanup(user) {
+  const cloned = _.clone(user);
+  delete cloned.password;
+  return cloned;
+}
+
+function setUserPassword(user, pwd, cb) {
+  hash.saltAndHash(pwd, function(err, hashed) {
+    if (err) {
+      return cb(err);
+    }
+    user.password = hashed;
+    user.passwordAttempts = 0;
+    cb(null, user);
+  });
+}
+
+/**
+ * Helper function to reset underlying collection for testing
+ */
+exports.setAll = function(arr) {
+  users = arr;
+};
+
+exports.all = function(cb) {
+  return cb(null, _.map(users, cloneAndCleanup));
+};
+
+exports.read = function(id, cb) {
+  const user = findById(id);
+  if (!user) {
+    return cb(new Error('User not found'));
+  }
+  cb(null, cloneAndCleanup(user));
+};
+
+exports.byUsername = function(username, cb) {
+  const user = findByUsername(username);
+  if (!user) {
+    return cb(new Error('User not found'));
+  }
+  cb(null, cloneAndCleanup(user));
+};
+
+
+exports.create = function(user, cb) {
+  user = _.clone(user);
+  user.id = shortid.generate();
+  setUserPassword(user, user.password, function(err, user) {
+    if (err) {
+      return cb(err);
+    }
+    users.push(user);
+    cb(null, cloneAndCleanup(user));
+  });
+};
+
+exports.update = function(id, user, cb) {
+  const originalUser = findById(id);
+  // avoid updating the password and id on this method
+  delete user.id;
+  delete user.password;
+  _.assign(originalUser, user);
+  if (!originalUser) {
+    return cb(new Error('User not found'));
+  }
+  cb(null, cloneAndCleanup(originalUser));
+};
+
+exports.updatePassword = function(username, oldPwd, newPwd, cb) {
+  const user = findByUsername(username);
+  if (!user) {
+    return cb(new Error('User not found'));
+  }
+  hash.verify(oldPwd, user.password, function(err, match) {
+    if (err || !match) {
+      return cb(new Error('Old password invalid'));
+    }
+    setUserPassword(user, newPwd, function(err, user) {
+      if (err) {
+        return cb(err);
+      }
+      cb(null, cloneAndCleanup(user));
+    });
+  });
+};
+
+function calculateDelay(attempts) {
+  return (Math.pow(2, attempts) - 1) * ATTEMPT_DELAY_MS;
+}
+
+const verifyErrMessage = 'User not found with supplied credentials';
+
+function doVerify(password, user, cb) {
+  // user.password here should be the stored hash
+  hash.verify(password, user.password, function(err, match) {
+    if (err) {
+      return cb(err);
+    }
+    if (!match) {
+      user.passwordAttempts = user.passwordAttempts || 0;
+      user.passwordAttempts++;
+      cb(new Error(verifyErrMessage), false);
+    } else {
+      user.passwordAttempts = 0;
+      cb(null, true);
+    }
+  });
+}
+
+exports.verifyPassword = function(username, password, cb) {
+  const user = findByUsername(username);
+  if (!user) {
+    return cb(new Error());
```

Comment:
Error message for both username not found or wrong password should be the same

Ref: https://www.owasp.org/index.php/Authentication_Cheat_Sheet#Authentication_and_Error_Messages